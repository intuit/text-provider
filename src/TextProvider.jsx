import React from 'react';
import PropTypes from 'prop-types';

export const TextContext = React.createContext({});

const TextProvider = (props) => {
  const { globalText, children } = props;
  return <TextContext.Provider value={globalText}>{children}</TextContext.Provider>;
};

TextProvider.propTypes = {
  globalText: PropTypes.objectOf(PropTypes.string),
  children: PropTypes.node.isRequired,
};

TextProvider.defaultProps = {
  globalText: {},
};

export { TextProvider as default };
