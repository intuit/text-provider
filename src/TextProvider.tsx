import React from 'react';

export const TextContext: React.Context<{}> = React.createContext({});

/**
 * Definition
 */
interface TextProviderProps {

/**
 * Key-value pair of text strings
 */
    globalText?: { [key: string]: string};

    children: any;
}

const TextProvider: React.FC<TextProviderProps> = (props: TextProviderProps) => {
  const { globalText, children } = props;
  return (
    <TextContext.Provider value={globalText}>
      {children}
    </TextContext.Provider>
  );
};

/**
 * Set default props
 */
TextProvider.defaultProps = {
  globalText: {},
};

export { TextProvider as default };
