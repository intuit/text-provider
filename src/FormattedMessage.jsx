import React from 'react';
import PropTypes from 'prop-types';
import { TextContext } from './TextProvider';

const FormattedMessage = (props) => (
  <TextContext.Consumer>
    {(context) => {
      /**
       * We get the text to be injected from react's context. Here its TextContext
       */
      const {
        id, values, alt, ariaLabel, automationId,
      } = props;
      /* eslint-disable-next-line react/destructuring-assignment */
      let messageString = Object.prototype.hasOwnProperty.call(context, id) ? context[id] : alt;
      /**
       * Iterate through all the keys given as the prop and replace with corresponding values.
       */
      Object.keys(values).forEach((key) => {
        messageString = messageString.replace(`{${key}}`, values[key]);
      });
      /* Do not frown over this, many a times you need to do this, like :
       * https://github.com/yahoo/react-intl/blob/master/src/components/html-message.js#L86
       */
      const style = {
        fontSize: 'inherit',
      };

      /* eslint-disable react/no-danger */
      return (
        <span
          automation-id={(automationId === '') ? id : automationId}
          aria-label={(ariaLabel === '') ? messageString : ariaLabel}
          style={style}
          dangerouslySetInnerHTML={{ __html: messageString }}
        />
      );
    }}
  </TextContext.Consumer>
);

FormattedMessage.propTypes = {
  id: PropTypes.string.isRequired,
  values: PropTypes.objectOf(PropTypes.object),
  alt: PropTypes.string,
  ariaLabel: PropTypes.string,
  automationId: PropTypes.string,
};

FormattedMessage.defaultProps = {
  values: {},
  alt: '',
  ariaLabel: '',
  automationId: '',
};

export { FormattedMessage as default };
