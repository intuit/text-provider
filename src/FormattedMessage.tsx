import React from 'react';
import { TextContext } from './TextProvider';


interface FormattedMessageProps {
  id: string;

  values?: { [name:string]: string};

  alt?: string;

}

const FormattedMessage: React.FC<FormattedMessageProps> = (props: FormattedMessageProps) => (
  <TextContext.Consumer>
    {(context: { [key: string]: string}) => {
      /**
       * We get the text to be injected from react's context. Here its TextContext
       */
      const { id, values, alt } = props;

      /* eslint-disable-next-line react/destructuring-assignment */
      let messageString = Object.prototype.hasOwnProperty.call(context, id) ? context[id] : alt;
      /**
       * Iterate through all the keys given as the prop and replace with corresponding values.
       */
      Object.keys(values).forEach((key: string) => {
        messageString = messageString.replace(`{${key}}`, values[key]);
      });
      /* Do not frown over this, many a times you need to do this, like :
       * https://github.com/yahoo/react-intl/blob/master/src/components/html-message.js#L86
       */
      const style = {
        fontSize: 'inherit',
      };
      /* eslint-disable react/no-danger */
      return <span style={style} dangerouslySetInnerHTML={{ __html: messageString }} />;
    }}
  </TextContext.Consumer>
);

FormattedMessage.defaultProps = {
  values: {},
  alt: '',
};

export { FormattedMessage as default };
