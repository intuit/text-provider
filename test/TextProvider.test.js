import React from 'react';
import renderer from 'react-test-renderer';
import TextProvider from '../src/TextProvider';
import FormattedMessage from '../src/FormattedMessage';

describe('<TextProvider /> rendering', () => {
  const randomId = 'Random Id';
  const values = {
    valueToBeInjected: 'Random Value',
  };
  const sampleText = {
    'Random Id': '{valueToBeInjected}',
  };

  it('should have default globalText prop', () => {
    expect(TextProvider.defaultProps.globalText).toBeDefined();
  });

  it('should render correctly', () => {
    let wrapper = renderer.create(
      <TextProvider
        globalText={sampleText}
        children={<FormattedMessage id={randomId} values={values} />}
      />,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should replace the id with value in the context message', () => {
    let wrapper = renderer.create(
      <TextProvider
        globalText={sampleText}
        children={<FormattedMessage id={randomId} values={values} />}
      />,
    );
    expect(wrapper.toJSON().props.style.fontSize).toBe('inherit');
    expect(wrapper.toJSON().props.dangerouslySetInnerHTML.__html).toBe(values['valueToBeInjected']);
  });
});
