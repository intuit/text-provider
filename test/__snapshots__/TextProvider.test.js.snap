// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextProvider /> rendering should render correctly 1`] = `
<span
  dangerouslySetInnerHTML={
    Object {
      "__html": "Random Value",
    }
  }
  style={
    Object {
      "fontSize": "inherit",
    }
  }
/>
`;
